features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/net/RumRequestFactory.kt

```diff
@@ -73,40 +86,90 @@ internal class RumRequestFactory(
 
     private fun buildHeaders(
         requestId: String,
-        clientToken: String,
-        source: String,
-        sdkVersion: String
+        idempotencyKey: String?,
+        context: DatadogContext
     ): Map<String, String> {
-        return mapOf(
-            RequestFactory.HEADER_API_KEY to clientToken,
-            RequestFactory.HEADER_EVP_ORIGIN to source,
-            RequestFactory.HEADER_EVP_ORIGIN_VERSION to sdkVersion,
+        val headers = mutableMapOf(
+            RequestFactory.HEADER_API_KEY to context.clientToken,
+            RequestFactory.HEADER_EVP_ORIGIN to context.source,
+            RequestFactory.HEADER_EVP_ORIGIN_VERSION to context.sdkVersion,
             RequestFactory.HEADER_REQUEST_ID to requestId
         )
+        if (idempotencyKey != null) {
+            headers[RequestFactory.DD_IDEMPOTENCY_KEY] = idempotencyKey
+        }
+        return headers
```

Comment:
yup I also had this in mind but I was thinking that might be a bit overkill...I could actually do it differently without having 2 create 2 maps. Let me change this.

Reply:
Actually the reason why I did this was that I return a `immutable map`. If you call `toMap` under the hood it also creates an `LinkedHashMap` so still mutable. So I think I have an uglier way but more optimised.

Reply:
So basically I will leave it like this because under the hood after a quick research `toMap` returns like I mentioned above a `LinkedHashMap`. Actually I would have expected to return an `ImmutableMap` but apparently the immutability is only at the type (surface level). So in a Kotlin code you could easily do this:
```
  val map = mapOf(...)
  val immutableMap = map.toMap()
  val hashMap = immutableMap as LinkedHashMap
```